### PR TITLE
Use reservoir sampling to select one host from priority list

### PR DESCRIPTION
**Pull request by hainesc**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
> /kind api-change
> /kind bug

/kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:

The function `selectHost` was designed to select one host from priority list with a highest score, However, the current manner use round-robin and it has to keep a var named `lastNodeIndex` in the struct of `genericScheduler`, and it introduces a helper function `findMaxScores`, both are not needed if we use reservoir sampling, which is more suitable for selecting one candidate with same possibility from a set, and the code will be cleaner. For the detail of reservoir sampling, see [here](https://en.wikipedia.org/wiki/Reservoir_sampling).

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/scheduler/core/generic_scheduler.go [hunk 1]
```diff
@@ -264,34 +264,29 @@ func (g *genericScheduler) Predicates() map[string]predicates.FitPredicate {
...
 			maxScore = hp.Score
-			maxScoreIndexes = maxScoreIndexes[:0]
-			maxScoreIndexes = append(maxScoreIndexes, i)
+			ix = i
+			cntOfMaxScore = 1
 		} else if hp.Score == maxScore {
-			maxScoreIndexes = append(maxScoreIndexes, i)
+			cntOfMaxScore++
+			if rand.Intn(cntOfMaxScore) == 0 {
+				// Replace the candidate with possibility of 1/cntOfMaxScore
```
Comment (ahg-g) [inline pkg/scheduler/core/generic_scheduler.go]: s/possibility of/probability

**Review by ahg-g — COMMENTED** [review]
This is nice!

File: pkg/scheduler/core/generic_scheduler.go (hunk 1 above)
Comment (hainesc) [reply pkg/scheduler/core/generic_scheduler.go]: > s/possibility of/probability

done

**Comment by hainesc** [comment]
ping @resouer @misterikkit for review and ok-to-test label

Thanks.

**Comment by hainesc** [comment]
ping @kubernetes/sig-scheduling-pr-reviews for review

**Review by yastij — APPROVED** [review]
/lgtm
/priority backlog
/assign @misterikkit

**Comment by yastij** [comment]
@hainesc - please also take a look at the tests. some of them rely on the ordering, introducing a rand can make it harder to test.

**Comment by yastij** [comment]
@bsalamat - while i can buy the logic behind the change, I'm afraid that this might over-complicate our testing logic, thoughts ?

**Comment by hainesc** [comment]
> @hainesc - please also take a look at the tests. some of them rely on the ordering, introducing a rand can make it harder to test.

I will fix the failing tests. I think it should not rely on the ordering.

**Comment by hainesc** [comment]
ping @yastij for `ok-to-test`

**Comment by hainesc** [comment]
The failing test `pull-kubernetes-kubemark-e2e-gce-big` may be due to https://github.com/kubernetes/test-infra/issues/12778, I saw the same error output. The flaky test has blocked several PRs, I will keep a close watch on that.

**Comment by hainesc** [comment]
All tests success and none of them rely ordering. Awesome test cases!
ping @yastij May you PTAL whether the pr is OK when you are free.

Thanks!

**Comment by ahg-g** [comment]
Can you please run https://github.com/kubernetes/kubernetes/tree/master/test/integration/scheduler_perf with and without this change just to make sure we are not introducing performance regression.

**Comment by hainesc** [comment]
> Can you please run https://github.com/kubernetes/kubernetes/tree/master/test/integration/scheduler_perf with and without this change just to make sure we are not introducing performance regression.

Good advise. I did some similar. I wrote a mock scheduler with functions `selectHostRoundRobin` and `selectHostReservoir` implemented as their names, then run a bench test on 100 nodes with random(rand.Intn(20)) score. It shows selectHostReservoir wins.
```go
// scheduler.go
package scheduler
import (
	"fmt"
	"math/rand"
)

type MockScheduler struct {
	lastNodeIndex uint64
	r *rand.Rand
}

type Priority struct {
	Host string
	Score int
}

type PriorityList []Priority

func (m *MockScheduler) selectHostRoundRobin(priorityList PriorityList) (string, error) {
	if len(priorityList) == 0 {
		return "", fmt.Errorf("empty")
	} 
	maxScores := findMaxScores(priorityList)
	ix := int(m.lastNodeIndex % uint64(len(maxScores)))
	m.lastNodeIndex++
	return priorityList[maxScores[ix]].Host, nil
}

func findMaxScores(priorityList PriorityList) []int {
	maxScoreIndexes := make([]int, 0, len(priorityList)/2)
	maxScore := priorityList[0].Score
	for i, hp := range priorityList {
		if hp.Score > maxScore {
			maxScore = hp.Score
			maxScoreIndexes = maxScoreIndexes[:0]
			maxScoreIndexes = append(maxScoreIndexes, i)
		} else if hp.Score == maxScore {
			maxScoreIndexes = append(maxScoreIndexes, i)
		}
	}
	return maxScoreIndexes
}

func (m *MockScheduler) selectHostReservoir(priorityList PriorityList) (string, error) {
	if len(priorityList) == 0 {
		return "", fmt.Errorf("empty")
	}
	maxScore := priorityList[0].Score
	selected := priorityList[0].Host
	cntOfMaxScore := 1
	for _, hp := range priorityList[1:] {
		if hp.Score > maxScore {
			maxScore = hp.Score
			selected = hp.Host
			cntOfMaxScore = 1
		} else if hp.Score == maxScore {
			cntOfMaxScore++
			if (rand.Intn(cntOfMaxScore) == 0) {
				selected = hp.Host
			}
		}
	}
	return selected, nil
}
```

```go
// scheduler_test.go
package scheduler

import (
	"testing"
	"math/rand"
	"time"
	"strconv"
)
func BenchmarkSelectHostRoundRobin(b *testing.B) {
	rand.Seed(time.Now().UnixNano())
	scheduler := &MockScheduler{}
	nodes := make([]Priority, 100, 100)
	for i := 0; i < 100; i++ {
		nodes[i].Host = "node" + strconv.Itoa(i)
	}
	for i := 0; i < 100; i++ {
		nodes[i].Score = rand.Intn(20)
	}
	b.ResetTimer()
	for n := 0; n < b.N; n++ {
		_, _ = scheduler.selectHostRoundRobin(nodes)
	}
}

func BenchmarkSelectHostReservoir(b *testing.B) {
	rand.Seed(time.Now().UnixNano())
	scheduler := &MockScheduler{}
	nodes := make([]Priority, 100, 100)
	for i := 0; i < 100; i++ {
		nodes[i].Host = "node" + strconv.Itoa(i)
	}
	for i := 0; i < 100; i++ {
		nodes[i].Score = rand.Intn(20)
	}
	b.ResetTimer()
	for n := 0; n < b.N; n++ {
		_, _ = scheduler.selectHostReservoir(nodes)
	}
}
```

Result:
```
$ go test -bench=.
goos: darwin
goarch: amd64
pkg: github.com/hainesc/scheduler/bench
BenchmarkSelectHostRoundRobin-8   	 3000000	       394 ns/op
BenchmarkSelectHostReservoir-8    	 5000000	       218 ns/op
PASS
ok  	github.com/hainesc/scheduler/bench	3.050s
$ go test -bench=.
goos: darwin
goarch: amd64
pkg: github.com/hainesc/scheduler/bench
BenchmarkSelectHostRoundRobin-8   	 5000000	       416 ns/op
BenchmarkSelectHostReservoir-8    	 5000000	       218 ns/op
PASS
ok  	github.com/hainesc/scheduler/bench	4.712s
```

**Comment by ahg-g** [comment]
Nice! did you try the one I linked to?

**Comment by hainesc** [comment]
> Nice! did you try the one I linked to?

I am trying, but it is time consuming to configure my dev env before it run.

**Comment by hainesc** [comment]
> Nice! did you try the one I linked to?

Hi, @ahg-g, So many output lines. Which data is your need?

**Comment by hainesc** [comment]
ping @ahg-g PTAL

**Comment by hainesc** [comment]
> /lgtm

Thanks! ^_^

**Comment by yastij** [comment]
re-applying lgtm per my first comment

/lgtm

**Comment by hainesc** [comment]
ping @misterikkit for approve, please take a look. Thanks.

**Comment by hainesc** [comment]
ping @yastij Can you help me ping an approver? It is a long time since the `lgtm` label. I don't know who is responsible for this pr.

**Comment by ahg-g** [comment]
we are in code freeze now, even if it gets approved, it is not going to be merged until the freeze is over (probably mid June).

**Comment by hainesc** [comment]
Hi @ahg-g, Can this pr get merged?

**Comment by hainesc** [comment]
ping @yastij for approve. 
Thanks

**Comment by gaorong** [comment]
thanks @hainesc  , I love this creative solution, but it seems this will break the previous round-robin manner to select a node and change to random selecting?

**Comment by hainesc** [comment]
> it seems this will break the previous round-robin manner to select a node and change to random selecting?

@gaorong Yes, as shown by the benchmark, random selecting wins a lot against the round-robin manner in performance. And also, as you see, the code is more cleaner than the previous.

I'd like to know what's your concern when we break the round-robin? In my opinion, there should be no assumption which node(with the highest score) will be selected, the only expected thing is that the candidate should be equal in probability.

**Comment by misterikkit** [comment]
@gaorong The current "round robin" behavior doesn't actually result in round robin scheduling. Rather, the `lastNodeIndex` is used to index into the slice of nodes which have the highest ranking. The size of this slice is likely to be different for each pod we schedule.

@bsalamat can confirm, but the intent of this mechanic is to prevent one node from getting all the pending pods when there are other equivalent nodes.

**Comment by hainesc** [comment]
Thanks @misterikkit Yes, the input slice is likely to be different in scores and order as the cluster is dynamic.

Looking forward this PR getting merged (yeah, a long time wait), as it is proved to be better than the current code. It is approved by @ahg-g and @yastij 

Thanks, all!

**Comment by gaorong** [comment]
> the intent of this mechanic is to prevent one node from getting all the pending pods when there are other equivalent nodes.

thanks misterikkit@,  after rethinking, I guess the `lastNodeIndex` should keep as it is. although it can not guarantee round-robin,  it can try to provide more fairly selecting among equivalent nodes than random selecting.

**Comment by ahg-g** [comment]
@Huang-Wei @bsalamat I like this PR, it allows removing a member variable of genericScheduler without compromising semantics (the minimal perf improvement is a cherry on the top).

**Comment by misterikkit** [comment]
> it can try to provide fairer selecting among equivalent nodes than random selecting.

@gaorong, I took a deeper look into the code, and the filter step makes no promises about stable ordering. In fact, it evaluates nodes in [parallel][1]. So in that sense, we are already doing an arbitrary, undefined distribution with `lastNodeIndex`. Switching to a uniform distribution ought to increase fairness. (It is unlikely to make it noticeably worse.)

[1]: https://github.com/kubernetes/kubernetes/blob/a9a68c5e4439423617722cae33da20204f186df4/pkg/scheduler/core/generic_scheduler.go#L484-L531

**Comment by Huang-Wei** [comment]
Thanks @misterikkit for the above comment. Simply put, we are doing (1) nodes shuffling + (2) (to some extent) fair nodes pickup by `lastIndex` which still results in an _eventual shuffling_.
So IMO it's ok to onboard PR (which is shuffling*2).

**Comment by gaorong** [comment]
thank you guys for clarification.

**Review by Huang-Wei — COMMENTED** [review]
/lgtm
/approve

Thanks, @hainesc !

**Review by yb01 — COMMENTED** [review]
/lgtm.
this change might node improve perf but it simplify the code and randomized the node selection. so let's take it.

**Pull request merged**